Raising proposal based on PR discussion for: Enhance KafkaBridge resurce with consumer inactivity timeout and HTTP consumer/producer parts enablement (#111)
strimzi/proposals: 071-kafkabrige-consumer-producer.md

## Proposal

This proposal suggests based on discussions here: https://github.com/strimzi/strimzi-kafka-operator/pull/9820 to add the `http enablement` for the `consumer` and `producer` in their respective sessions, i.e. `spec.consumer.enabled` and `spec.producer.enabled`, as well as adding `timeoutSeconds` to the `spec.consumer` properties.

Suggestion:

```yaml
apiVersion: "kafka.strimzi.io/v1beta2"
kind: "KafkaBridge"
metadata:
  name: "test-kafka-bridge"
spec:
  replicas: 1
  image: "my-test-image"
  bootstrapServers: "my-cluster-kafka:9092"
  consumer:
    enabled: true
    timeoutSeconds: 60
    config:
      foo: "bar"
  producer:
    enabled: false
    config:
      foo: "buz"
  enableMetrics: false
```

## Affected/not affected projects

Strimzi Kafka Bridge - PR raised for the HTTP Bridge to use the new ENV parameters introduced from the operator: https://github.com/strimzi/strimzi-kafka-bridge/pull/882

## Compatibility

`http.consumer.enabled` and `http.producer.enabled` both defaults to `true`, thus there are no impacts on backward compatibility.

> 

Reviewer: for thoroughness, we should talk here about compatibility aspects of `timeoutSeconds`.  If field is omitted from CR, I would expect the effect on the bridge would be as if `http.consumer.timeoutSeconds` had the value [`-1`](https://github.com/strimzi/strimzi-kafka-bridge/blob/25b4106bd5cf3b9b4c8f128b6c284159bb736149/src/main/java/io/strimzi/kafka/bridge/http/HttpConfig.java#L50).

Reviewer 2: Yes when you omit `http.consumer.timeoutSeconds` from the application.properties, it takes `-1` as default value (no timeout). Missing `timeoutSeconds` in the CR would mean not setting it in the application.properties so providing the same effect. Of course, I agree it should be clarified in this section.

# Enhance KafkaBridge resource with consumer inactivity timeout and HTTP consumer/producer parts enablement

Providing support in the Strimzi Kubernetes Operator for following properties supported in the Strimzi HTTP Kafka Bridge:
* `http.timeoutSeconds`   - For deleting inactive consumers after a timeout (disabled by default).
* `http.consumer.enabled` - To enable/disable the HTTP consumer part (enabled by default).
* `http.producer.enabled` - To enable/disable the HTTP producer part (enabled by default).

## Current situation

Properties are not yet supported in the Strimzi Kubernetes Operator.

## Motivation

Raised in the discussion here: https://github.com/strimzi/strimzi-kafka-operator/issues/8732 and triaged on 29.6.2023: We should enable the configuration of these fields. But how should the API look like? Should have a proposal to clarify the API changes.

## Proposal

> This proposal suggests based on discussions here: https://github.com/strimzi/strimzi-kafka-operator/pull/9820 to add the `http enablement` for the `consumer` and `producer` in their respective sessions, i.e. `spec.consumer.enabled` and `spec.producer.enabled`, as well as adding `timeoutSeconds` to the `http` properties.

Reviewer: This sentence is now inconsistent with the updated API proposal.

Reviewer 2: +1 for @k-wall 
Also I would avoid the code formatting on "http enablement", "consumer" and "producer" but using it for actual code and fields.

Suggestion:

```yaml
apiVersion: "kafka.strimzi.io/v1beta2"
kind: "KafkaBridge"
metadata:
  name: "test-kafka-bridge"
spec:
  replicas: 1
  image: "my-test-image"
  bootstrapServers: "my-cluster-kafka:9092"
  consumer:
    enabled: true
    timeoutSeconds: 60
    config:
      foo: "bar"
  producer:
    enabled: false

>     timeoutSeconds: 60

Reviewer: I don't think producer actually has the concept of timeout.  It is only the consumer that does.

Reviewer 2: +1 It has not, you are right.

    config:
      foo: "buz"
  enableMetrics: false
```

## Affected/not affected projects

Strimzi Kafka Bridge - PR raised for the HTTP Bridge to use the new ENV parameters introduced from the operator: https://github.com/strimzi/strimzi-kafka-bridge/pull/882

## Compatibility

`http.consumer.enabled` and `http.producer.enabled` both defaults to `true`, thus there are no impacts on backward compatibility.

## Rejected alternatives

### 1.
Initially in the implementation suggestion: https://github.com/strimzi/strimzi-kafka-operator/pull/9820, the `enabled` properties for `producer` and `consumer` were added to as part of the `http` properties, i.e. `spec.http.consumer.enabled` and `spec.http.producer.enabled`.

Alternative suggestion:

```yaml
apiVersion: "kafka.strimzi.io/v1beta2"
kind: "KafkaBridge"
metadata:
  name: "test-kafka-bridge"
spec:
  replicas: 1
  image: "my-test-image"
  bootstrapServers: "my-cluster-kafka:9092"
  http:
    timeoutSeconds: 60
    producer:
      enabled: false
    consumer:
      enabled: true
  consumer:
    config:
      foo: "bar"
  producer:
    config:
      foo: "buz"
  enableMetrics: false
```

### 2.

> `timeoutSeconds` has been moved from `http` to `consumer` and `producer` section.

Reviewer 2: again `timeoutSeconds` is valid for consumer only.

# Enhance KafkaBridge resource with consumer inactivity timeout and HTTP consumer/producer parts enablement

Providing support in the Strimzi Kubernetes Operator for following properties supported in the Strimzi HTTP Kafka Bridge:

> * http.consumer.timeoutSeconds   - For deleting inactive consumers after a timeout (disabled by default).

Reviewer 2: ```suggestion
* `http.consumer.timeoutSeconds`   - For deleting inactive consumers after a timeout (disabled by default).
```

> * http.consumer.enabled          - To enable/disable the HTTP consumer part (enabled by default).

Reviewer 2: ```suggestion
* `http.consumer.enabled`          - To enable/disable the HTTP consumer part (enabled by default).
```

> * http.producer.enabled          - To enable/disable the HTTP producer part (enabled by default).

Reviewer 2: ```suggestion
* `http.producer.enabled`          - To enable/disable the HTTP producer part (enabled by default).
```

## Current situation

Properties are not yet supported in the Strimzi Kubernetes Operator.

## Motivation

Raised in the discussion here: https://github.com/strimzi/strimzi-kafka-operator/issues/8732 and triaged on 29.6.2023: We should enable the configuration of these fields. But how should the API look like? Should have a proposal to clarify the API changes.

## Proposal

> This proposal suggests based on discussions here: https://github.com/strimzi/strimzi-kafka-operator/pull/9820 to add the http enablement for the consumer and producer in their respective sessions, i.e. spec.consumer.enabled and spec.producer.enabled, as well as adding timeoutSeconds to the spec.consumer properties.

Reviewer 2: ```suggestion
This proposal suggests based on discussions here: https://github.com/strimzi/strimzi-kafka-operator/pull/9820 to add the http enablement for the consumer and producer in their respective sessions, i.e. `spec.consumer.enabled` and `spec.producer.enabled`, as well as `spec.consumer.timeoutSeconds` property.
```

Suggestion:

```yaml
apiVersion: "kafka.strimzi.io/v1beta2"
kind: "KafkaBridge"
metadata:
  name: "test-kafka-bridge"
spec:
  replicas: 1
  image: "my-test-image"
  bootstrapServers: "my-cluster-kafka:9092"
  consumer:
    enabled: true
    timeoutSeconds: 60
    config:
      foo: "bar"
  producer:
    enabled: false
    config:
      foo: "buz"
  enableMetrics: false
```

## Affected/not affected projects

Strimzi Kafka Bridge - PR raised for the HTTP Bridge to use the new ENV parameters introduced from the operator: https://github.com/strimzi/strimzi-kafka-bridge/pull/882

## Compatibility

> http.consumer.enabled and http.producer.enabled both defaults to true, thus there are no impacts on backward compatibility.

Reviewer 2: ```suggestion
Not specifying `spec.consumer.enabled` or `spec.producer.enabled` implies not configuring the corresponding bridge parameters `http.consumer.enabled` and `http.producer.enabled` which default to true, thus there are no impacts on backward compatibility.
```

> If http.consumer.timeoutSeconds is not specified, it should default to -1 to have the same effect, as when it is omitted from the application.properties.

Reviewer 2: It seems you are referring to the property you would have in the `KafkaBridge` which is omitted but it's not `http.consumer.timeoutSeconds` but it's `spec.consumer.timeoutSeconds`.

```suggestion
Not specifying `spec.consumer.timeoutSeconds` implies not configuring the corresponding bridge parameter `http.consumer.timeoutSeconds` which defaults to -1 to have the same effect (no timeout), thus there are no impacts on backward compatibility.
```

## Rejected alternatives

### 1.

> Initially in the implementation suggestion: https://github.com/strimzi/strimzi-kafka-operator/pull/9820, the enabled properties for producer and consumer were added to as part of the http properties, i.e. spec.http.consumer.enabled and spec.http.producer.enabled.

Reviewer 2: ```suggestion
Initially in the implementation suggestion: https://github.com/strimzi/strimzi-kafka-operator/pull/9820, the enabled properties for producer and consumer were added to as part of the http properties, i.e. `spec.http.consumer.enabled` and `spec.http.producer.enabled`.
```

Alternative suggestion:

```yaml
apiVersion: "kafka.strimzi.io/v1beta2"
kind: "KafkaBridge"
metadata:
  name: "test-kafka-bridge"
spec:
  replicas: 1
  image: "my-test-image"
  bootstrapServers: "my-cluster-kafka:9092"
  http:
    timeoutSeconds: 60
    producer:
      enabled: false
    consumer:
      enabled: true
  consumer:
    config:
      foo: "bar"
  producer:
    config:
      foo: "buz"
  enableMetrics: false
```

### 2.

> timeoutSeconds has been moved from http to consumer section.

Reviewer 2: I think we can remove this sentece. It's just part of the proposed solution.
